### PR TITLE
BoundedContext factory

### PR DESCRIPTION
**Pull request by dmdashenkov**
In this PR a new component takes the scene - `Contexts`.

`Contexts` is a factory for the `BoundedContext` instances based on GAE Datastore.

By introducing this component we resolve two problems:
 - fix the `TenantIndex` inconsistency between the `BoundedContext` and `DatastoreTenants` for the single-tenant applications (or BCs);
 - resolve the _rely-on-user_ issue when building the multitenant `BoundedContext`s.

### CODE REVIEW

**Comment by dmdashenkov** [comment]
@armiol, PTAL.

File: gcd/src/main/java/org/spine3/server/datastore/package-info.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * This package contains components for working with the aspects of using the Google Cloud Datastore
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/datastore/package-info.java]: ` ... of Google Cloud Datastore usage`

File: gcd/src/main/java/org/spine3/server/datastore/Contexts.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+     *
+     * <p>For the most applications this config will be convenient if the passed
+     * {@link DatastoreStorageFactory} is configured properly. Though, the successive calls to
+     * the corresponding {@link BoundedContext.Builder Builder} methods may override this
+     * configuration.
+     *
+     * @param storageFactory the {@link StorageFactory} to use in the result {@link BoundedContext}
+     * @return new instance of the {@link BoundedContext.Builder} with the specified parameters
+     */
+    public static BoundedContext.Builder from(DatastoreStorageFactory storageFactory) {
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/datastore/Contexts.java]: `onTopOf(DatastoreStorageFactory storageFactory)`

File: gcd/src/main/java/org/spine3/server/datastore/Contexts.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+     * {@link DatastoreStorageFactory}.
+     *
+     * <p>The result builder will have the next parameters set:
+     * <ul>
+     *     <li>{@link BoundedContext.Builder#setStorageFactorySupplier(Supplier) StorageFactory};
+     *     <li>{@link BoundedContext.Builder#setTenantIndex(TenantIndex) TenantIndex};
+     *     <li>{@linkplain BoundedContext.Builder#setMultitenant(boolean) multitenancy}.
+     * </ul>
+     *
+     * <p>For the most applications this config will be convenient if the passed
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/datastore/Contexts.java]:
```
<p>In a majority of use cases the configuration of the produced {@link BoundedContext.Builder builder} is 
enough for operation. However, it is still possible to use the returned instance 
for further customization.
```

File: gcd/src/main/java/org/spine3/server/datastore/Contexts.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+    private Contexts() {
+        // Prevent initialization of this utility class
+    }
+
+    /**
+     * Creates new instance of the {@link BoundedContext.Builder} based of the passed
+     * {@link DatastoreStorageFactory}.
+     *
+     * <p>The result builder will have the next parameters set:
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/datastore/Contexts.java]:
```
<p> The returned instance has the following attributes pre-configured:
...
```

File: gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreStorageFactory.java [hunk 5]
```diff
@@ -213,6 +215,10 @@ public void close() throws Exception {
         // NOP
     }
 
+    /**
+     * @return an instance of a wrapper on the passed {@link Datastore}
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/storage/datastore/DatastoreStorageFactory.java]:
```
wrapper on
```
does not seem to be correct. Please review.

File: gcd/src/main/java/org/spine3/server/storage/datastore/tenant/DatastoreTenants.java [hunk 6]
```diff
@@ -39,6 +39,12 @@ private DatastoreTenants() {
     /**
      * Creates a {@link TenantIndex} for the given {@link Datastore}.
      *
+     * <p>It's highly recommended not to use this method directly when building a
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/storage/datastore/tenant/DatastoreTenants.java]:
```
<p>This method is intended for a manual {@code BoundedContext} configuration. To automate the setup routine please use {@link org.spine3.server.datastore.Contexts#onTopOf(...) Contexts.onTopOf()} method.
```

File: gcd/src/main/java/org/spine3/server/storage/datastore/tenant/DatastoreTenants.java [hunk 7]
```diff
@@ -70,10 +76,13 @@ private DatastoreTenants() {
      *
      * <p>A single-tenant app (or a single-tenant BoundedContext in a multitenant app) does not
      * require a {@code TenantIndex} to be set explicitly, so this method assumes that it is an a
-     * single tenant context.
+     * single tenant context. Though, not using this method when creating an instance of
```
Comment (armiol) [inline gcd/src/main/java/org/spine3/server/storage/datastore/tenant/DatastoreTenants.java]: Please revert this portion of changes. It's enough that we describe the details in `Contexts` and reference it above.

File: gcd/src/test/java/org/spine3/server/datastore/ContextsShould.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+        assertSame(factory, supplierOptional.get().get());
+        assertEquals(builder.isMultitenant(), factory.isMultitenant());
+        final Optional<? extends TenantIndex> tenantIndexOptional = builder.getTenantIndex();
+        assertTrue(tenantIndexOptional.isPresent());
+        assertThat(tenantIndexOptional.get(), instanceOf(TestNamespaceIndex.getType()));
+    }
+
+    private static DatastoreStorageFactory givenFactory() {
+        final DatastoreStorageFactory result = DatastoreStorageFactory.newBuilder()
+                .setDatastore(givenDatastore())
```
Comment (armiol) [inline gcd/src/test/java/org/spine3/server/datastore/ContextsShould.java]: The code isn't aligned.

**Review by armiol — CHANGES_REQUESTED** [review]
@dmdashenkov LGTM in general, but some comments needs to be addressed and reviewed once more before merge.

Also, please find and fix typos.

**Comment by dmdashenkov** [comment]
@armiol, PTAL again.

**Review by armiol — APPROVED** [review]
@dmdashenkov LGTM.

**Pull request merged**